Extend documentation regarding logging

jdksjolen opened the pull request:
Hi!

Here's some suggested improvements to the logging documentation. It's currently just a draft, but please feel welcome to come with any suggestions.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace

### Reviewers
 * [Jesper Wilhelmsson](https://openjdk.org/census#jwilhelm) (@JesperIRL - **Reviewer**) ⚠️ Review applies to [ab07b777](https://git.openjdk.org/guide/pull/91/files/ab07b777b19a02398d7d502c21a60e5c56def8cd)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - Author)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/guide pull/91/head:pull/91` \
`$ git checkout pull/91`

Update a local copy of the PR: \
`$ git checkout pull/91` \
`$ git pull https://git.openjdk.org/guide pull/91/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 91`

View PR using the GUI difftool: \
`$ git pr show -t 91`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/guide/pull/91.diff">https://git.openjdk.org/guide/pull/91.diff</a>

</details>

JesperIRL commented on src/guide/hotspot-development.md:
```diff
@@ -10,27 +10,76 @@
 
-The appropriate way to print logging output from HotSpot is through the [Unified Logging Framework (JEP 158)](https://openjdk.org/jeps/158). It gives you a lot of nice features and enables common command-line options for all logging. Messages can also be "decorated" with e.g. uptime, level, tags. The JEP contains a thorough description of the feature, but a quick example might look like:
+The appropriate way to print logging output from HotSpot is through the [Unified Logging Framework (JEP 158)](https://openjdk.org/jeps/158).
+It gives you a lot of nice features and enables common command-line options for all logging.
```
Please don't add line breaks in sections. Goes for sections below as well.

dholmes-ora replied:
Starting each sentence on a new line seems a very personal style choice - ref:
https://groups.google.com/g/pandoc-discuss/c/tEWDWL2mLBY?pli=1
It probably helps if also trying to maintain maximum line lengths, as editing may need fewer lines to be rewrapped. But personally I find it jarring to read.

jdksjolen replied:
Okay, this is probably a consequence of my editor being set up for coding and not for text editing then :-).

JesperIRL replied:
It's a personal style choice and it's the coding style used in the Developers' Guide. If lines are getting too long I recommend making your editor window narrower.

JesperIRL commented on src/guide/hotspot-development.md:
```diff
@@ -10,27 +10,76 @@
+The appropriate way to print logging output from HotSpot is through the [Unified Logging Framework (JEP 158)](https://openjdk.org/jeps/158).
+It gives you a lot of nice features and enables common command-line options for all logging.
+
+
```
One empty line is enough :-)

JesperIRL commented on src/guide/hotspot-development.md:
```diff
@@ -10,27 +10,76 @@
+It gives you a lot of nice features and enables common command-line options for all logging.
+
+
+A basic log message can be output like this.
```
Please end the line with a colon.

JesperIRL commented on src/guide/hotspot-development.md:
```diff
@@ -10,27 +10,76 @@
 
-The API should be similar to:
+Sometimes single line printf-style logging isn't enough.
+For example, it can be useful to group several log lines together or to use Hotspot's output stream API.
```
HotSpot should have a capitalized S.

JesperIRL commented on src/guide/hotspot-development.md:
```diff
@@ -10,27 +10,76 @@
+if(st.is_enabled()) {
+  st.print("I'm printing a lot of %s ", "arguments");
+  st.print("With a lot of extra info %d ", 3);
+  st.print_cr("and so it's useful to use a stream");
```
I think it would be useful to explain the difference between print and print_cr even though it may seem obvious to many. A single sentence just saying that print_cr will print a line break is enough.

JesperIRL commented on src/guide/hotspot-development.md:
```diff
@@ -10,27 +10,76 @@
+}
+~~~
+
+
```
One empty line is enough. The fact that you added two makes me think that you want to separate these two sections slightly. If so, use a subheader ##.

JesperIRL commented on src/guide/hotspot-development.md:
```diff
@@ -10,27 +10,76 @@
+~~~
+
+
+You turn on logging by starting the JVM with the `-Xlog` command line option specified.
```
I would phrase this slightly different: "You enable logging in the JVM by using the `-Xlog` command line option."

JesperIRL commented on src/guide/hotspot-development.md:
```diff
@@ -10,27 +10,76 @@
+
+
+You turn on logging by starting the JVM with the `-Xlog` command line option specified.
+For example, the messages from the examples would be visible if the JVM were run with any of the following flags:
```
"...from the examples above..."
"flags" -> "options"

JesperIRL commented on src/guide/hotspot-development.md:
```diff
@@ -10,27 +10,76 @@
 ~~~
 
-At the time of writing, the different log levels can be found in [src/hotspot/share/logging/log.hpp](https://github.com/openjdk/jdk/blob/master/src/hotspot/share/logging/log.hpp).
+A full description of the syntax of `-Xlog` is available in [JEP 158](https://openjdk.java.net/jeps/158), also mentioned above.
```
I would end the sentence directly after the link.

JesperIRL reviewed (changes requested):
Thank you for extending this section! This is good material that will be very helpful for anyone embarking on a VM hacking journey. I've made a few minor suggestions on style and language. On the technical content I trust you know that a lot better than me :-)

jdksjolen commented:
Hi Jesper!

Thank you for your review, I've addressed all except the one where I sent you a comment asking for clarification.

I also added some more content regarding `-Xlog`.

dholmes-ora commented on src/guide/hotspot-development.md:
```diff
@@ -10,27 +10,90 @@
 ~~~
 
-Where 'gc' and 'marking' are tags, and 'info' is the log level. This would be visible if the JVM were run with any of the following flags:
+Where 'gc' and 'marking' are tags, and 'info' is the log level. Tags associate log messages with certain subsystems or features and the log level determines the importance and verbosity of the message. The most verbose output is trace, and the least is error. The full list of tags is available [here](https://github.com/openjdk/jdk/blob/master/src/hotspot/share/logging/logTag.hpp) and the full lis ...
```
The full list of tags for a given VM is found from running `java -Xlog:help`.

jdksjolen replied:
Hi, I agree. I simply missed the comment when I updated the PR. I'll fix it :-).

dholmes-ora commented on src/guide/hotspot-development.md:
```diff
@@ -10,27 +10,90 @@
+
+~~~c++
+LogStream st(Log(gc, marking)::info());
+if(st.is_enabled()) {
```
Style nits: need space after `if` (here and elsewhere).

dholmes-ora commented on src/guide/hotspot-development.md:
```diff
@@ -10,27 +10,90 @@
+
+This specifies that:
+
+1. Log messages with info level and up with tags gc and marking to stdout.
+2. Log messages with warning level and up with tag alloc to stderr.
+3. Log messages with error level and up with tag breakpoint to file breakpoint.txt with the decorator level.
```
I suggest adding some commas to split each grouping e.g.

3. Log messages with error level and up, with tag breakpoint, to file breakpoint.txt, with the decorator level.

jdksjolen commented:
Hi,
I've gone through your comments and added fixes.

JesperIRL commented:
The current version looks good to me. I do agree with @dholmes-ora comment about mentioning `java -Xlog:help` on line 19. Even though it's mentioned at the end, there is no harm in mentioning it here as well. It's unlikely that the reader will read the entire section. The source code is harder to read that the help output.

jdksjolen commented:
I changed the reference of source files to `-Xlog:help` instead!

dholmes-ora reviewed (approved):
Looks good. Thanks

jdksjolen commented:
Thanks for all of the help!

/integrate

JesperIRL commented:
/sponsor
